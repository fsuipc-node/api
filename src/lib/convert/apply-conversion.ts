import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
