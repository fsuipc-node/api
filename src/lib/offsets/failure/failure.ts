import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const failure: OffsetList = {
  hydFailure: new Offset({
    value: 0xB62,
    name: 'hydFailure',
    category: OffsetCategory.FAILURE,
    description: 'hydraulic failure',
    type: Type.Byte,
    convert: '{VAL} ? 1 : 0',
    permission: 'w',
  }),
  brakeFailure: new Offset({
    value: 0xB63,
    name: 'brakeFailure',
    category: OffsetCategory.FAILURE,
    description: 'brake failure - left, right, total',
    convert: '{VAL} === `left` ? 0 : {VAL} === `right` ? 1 : 2',
    type: Type.Byte,
    permission: 'w',
  }),
  ADFFailure: new Offset({
    value: 0xB64,
    name: 'ADFFailure',
    category: OffsetCategory.FAILURE,
    description: 'ADF failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  ASIFailure: new Offset({
    value: 0xB65,
    name: 'ASIFailure',
    category: OffsetCategory.FAILURE,
    description: 'ASI failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  altimeterFailure: new Offset({
    value: 0xB66,
    name: 'altimeterFailure',
    category: OffsetCategory.FAILURE,
    description: 'altimeter failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  attitudeIndicatorFailure: new Offset({
    value: 0xB67,
    name: 'attitudeIndicatorFailure',
    category: OffsetCategory.FAILURE,
    description: 'attitude indicator failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  com1Failure: new Offset({
    value: 0xB68,
    name: 'com1Failure',
    category: OffsetCategory.FAILURE,
    description: 'COM1 failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  compassFailure: new Offset({
    value: 0xB69,
    name: 'compassFailure',
    category: OffsetCategory.FAILURE,
    description: 'compass failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  electricsFailure: new Offset({
    value: 0xB6A,
    name: 'electricsFailure',
    category: OffsetCategory.FAILURE,
    description: 'electrics failure',
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  engineFailure: new Offset({
    value: 0xB6B,
    name: 'engineFailure',
    category: OffsetCategory.FAILURE,
    description: 'engine failure',
    type: Type.BitArray,
    length: 4,
    permission: 'rw',
  }),
  fuelIndicatorFailure: new Offset({
    value: 0xB6C,
    name: 'fuelIndicatorFailure',
    category: OffsetCategory.FAILURE,
    description: 'fuel indicator failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  headingFailure: new Offset({
    value: 0xB6D,
    name: 'headingFailure',
    category: OffsetCategory.FAILURE,
    description: 'direction indicator failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  VSIFailure: new Offset({
    value: 0xB6E,
    name: 'VSIFailure',
    category: OffsetCategory.FAILURE,
    description: 'VSI failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  transponderFailure: new Offset({
    value: 0xB6F,
    name: 'transponderFailure',
    category: OffsetCategory.FAILURE,
    description: 'transponder failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  NAVFailure: new Offset({
    value: 0xB70,
    name: 'NAVFailure',
    category: OffsetCategory.FAILURE,
    description: 'NAV radio gauge failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  pitotFailure: new Offset({
    value: 0xB71,
    name: 'pitotFailure',
    category: OffsetCategory.FAILURE,
    description: 'pitot failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  turnCoordinatorFailure: new Offset({
    value: 0xB72,
    name: 'turnCoordinatorFailure',
    category: OffsetCategory.FAILURE,
    description: 'turn coordinator failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
  vacuumFailure: new Offset({
    value: 0xB73,
    name: 'vacuumFailure',
    category: OffsetCategory.FAILURE,
    description: 'vacuum failure',
    type: Type.Byte,
    convert: 'typeof {VAL} === `boolean` ? {VAL} ? 1 : 0 : !!{VAL}',
    permission: 'rw',
  }),
};
