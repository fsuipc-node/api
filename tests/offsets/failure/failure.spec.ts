import { failure as offsets } from '@offsets/failure/failure';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/failures', () => {
  const offsetsTestCases = [
    { name: 'hydFailure', value: true, expectedResult: 1 },
    { name: 'hydFailure', value: false, expectedResult: 0 },
    { name: 'brakeFailure', value: 'left', expectedResult: 0 },
    { name: 'brakeFailure', value: 'right', expectedResult: 1 },
    { name: 'brakeFailure', value: 'all', expectedResult: 2 },
    { name: 'brakeFailure', value: 'total', expectedResult: 2 },
    { name: 'ADFFailure', value: true, expectedResult: 1 },
    { name: 'ADFFailure', value: false, expectedResult: 0 },
    { name: 'ADFFailure', value: 1, expectedResult: true },
    { name: 'ADFFailure', value: 0, expectedResult: false },
    { name: 'ASIFailure', value: true, expectedResult: 1 },
    { name: 'ASIFailure', value: false, expectedResult: 0 },
    { name: 'ASIFailure', value: 1, expectedResult: true },
    { name: 'ASIFailure', value: 0, expectedResult: false },
    { name: 'altimeterFailure', value: true, expectedResult: 1 },
    { name: 'altimeterFailure', value: false, expectedResult: 0 },
    { name: 'altimeterFailure', value: 1, expectedResult: true },
    { name: 'altimeterFailure', value: 0, expectedResult: false },
    { name: 'attitudeIndicatorFailure', value: true, expectedResult: 1 },
    { name: 'attitudeIndicatorFailure', value: false, expectedResult: 0 },
    { name: 'attitudeIndicatorFailure', value: 1, expectedResult: true },
    { name: 'attitudeIndicatorFailure', value: 0, expectedResult: false },
    { name: 'com1Failure', value: true, expectedResult: 1 },
    { name: 'com1Failure', value: false, expectedResult: 0 },
    { name: 'com1Failure', value: 1, expectedResult: true },
    { name: 'com1Failure', value: 0, expectedResult: false },
    { name: 'compassFailure', value: true, expectedResult: 1 },
    { name: 'compassFailure', value: false, expectedResult: 0 },
    { name: 'compassFailure', value: 1, expectedResult: true },
    { name: 'compassFailure', value: 0, expectedResult: false },
    { name: 'electricsFailure', value: true, expectedResult: 1 },
    { name: 'electricsFailure', value: false, expectedResult: 0 },
    { name: 'electricsFailure', value: 1, expectedResult: true },
    { name: 'electricsFailure', value: 0, expectedResult: false },
    { name: 'fuelIndicatorFailure', value: true, expectedResult: 1 },
    { name: 'fuelIndicatorFailure', value: false, expectedResult: 0 },
    { name: 'fuelIndicatorFailure', value: 1, expectedResult: true },
    { name: 'fuelIndicatorFailure', value: 0, expectedResult: false },
    { name: 'headingFailure', value: true, expectedResult: 1 },
    { name: 'headingFailure', value: false, expectedResult: 0 },
    { name: 'headingFailure', value: 1, expectedResult: true },
    { name: 'headingFailure', value: 0, expectedResult: false },
    { name: 'VSIFailure', value: true, expectedResult: 1 },
    { name: 'VSIFailure', value: false, expectedResult: 0 },
    { name: 'VSIFailure', value: 1, expectedResult: true },
    { name: 'VSIFailure', value: 0, expectedResult: false },
    { name: 'transponderFailure', value: true, expectedResult: 1 },
    { name: 'transponderFailure', value: false, expectedResult: 0 },
    { name: 'transponderFailure', value: 1, expectedResult: true },
    { name: 'transponderFailure', value: 0, expectedResult: false },
    { name: 'NAVFailure', value: true, expectedResult: 1 },
    { name: 'NAVFailure', value: false, expectedResult: 0 },
    { name: 'NAVFailure', value: 1, expectedResult: true },
    { name: 'NAVFailure', value: 0, expectedResult: false },
    { name: 'pitotFailure', value: true, expectedResult: 1 },
    { name: 'pitotFailure', value: false, expectedResult: 0 },
    { name: 'pitotFailure', value: 1, expectedResult: true },
    { name: 'pitotFailure', value: 0, expectedResult: false },
    { name: 'turnCoordinatorFailure', value: true, expectedResult: 1 },
    { name: 'turnCoordinatorFailure', value: false, expectedResult: 0 },
    { name: 'turnCoordinatorFailure', value: 1, expectedResult: true },
    { name: 'turnCoordinatorFailure', value: 0, expectedResult: false },
    { name: 'vacuumFailure', value: true, expectedResult: 1 },
    { name: 'vacuumFailure', value: false, expectedResult: 0 },
    { name: 'vacuumFailure', value: 1, expectedResult: true },
    { name: 'vacuumFailure', value: 0, expectedResult: false },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
