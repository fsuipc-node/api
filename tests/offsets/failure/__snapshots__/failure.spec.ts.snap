// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/failures offsets list should have required properties 1`] = `
Object {
  "ADFFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "ADF failure",
    "length": undefined,
    "mapping": undefined,
    "name": "ADFFailure",
    "permission": "rw",
    "type": 0,
    "value": 2916,
  },
  "ASIFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "ASI failure",
    "length": undefined,
    "mapping": undefined,
    "name": "ASIFailure",
    "permission": "rw",
    "type": 0,
    "value": 2917,
  },
  "NAVFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "NAV radio gauge failure",
    "length": undefined,
    "mapping": undefined,
    "name": "NAVFailure",
    "permission": "rw",
    "type": 0,
    "value": 2928,
  },
  "VSIFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "VSI failure",
    "length": undefined,
    "mapping": undefined,
    "name": "VSIFailure",
    "permission": "rw",
    "type": 0,
    "value": 2926,
  },
  "altimeterFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "altimeter failure",
    "length": undefined,
    "mapping": undefined,
    "name": "altimeterFailure",
    "permission": "rw",
    "type": 0,
    "value": 2918,
  },
  "attitudeIndicatorFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "attitude indicator failure",
    "length": undefined,
    "mapping": undefined,
    "name": "attitudeIndicatorFailure",
    "permission": "rw",
    "type": 0,
    "value": 2919,
  },
  "brakeFailure": Offset {
    "category": "failure",
    "convert": "{VAL} === \`left\` ? 0 : {VAL} === \`right\` ? 1 : 2",
    "description": "brake failure - left, right, total",
    "length": undefined,
    "mapping": undefined,
    "name": "brakeFailure",
    "permission": "w",
    "type": 0,
    "value": 2915,
  },
  "com1Failure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "COM1 failure",
    "length": undefined,
    "mapping": undefined,
    "name": "com1Failure",
    "permission": "rw",
    "type": 0,
    "value": 2920,
  },
  "compassFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "compass failure",
    "length": undefined,
    "mapping": undefined,
    "name": "compassFailure",
    "permission": "rw",
    "type": 0,
    "value": 2921,
  },
  "electricsFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "electrics failure",
    "length": undefined,
    "mapping": undefined,
    "name": "electricsFailure",
    "permission": "rw",
    "type": 0,
    "value": 2922,
  },
  "engineFailure": Offset {
    "category": "failure",
    "convert": undefined,
    "description": "engine failure",
    "length": 4,
    "mapping": undefined,
    "name": "engineFailure",
    "permission": "rw",
    "type": 12,
    "value": 2923,
  },
  "fuelIndicatorFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "fuel indicator failure",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelIndicatorFailure",
    "permission": "rw",
    "type": 0,
    "value": 2924,
  },
  "headingFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "direction indicator failure",
    "length": undefined,
    "mapping": undefined,
    "name": "headingFailure",
    "permission": "rw",
    "type": 0,
    "value": 2925,
  },
  "hydFailure": Offset {
    "category": "failure",
    "convert": "{VAL} ? 1 : 0",
    "description": "hydraulic failure",
    "length": undefined,
    "mapping": undefined,
    "name": "hydFailure",
    "permission": "w",
    "type": 0,
    "value": 2914,
  },
  "pitotFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "pitot failure",
    "length": undefined,
    "mapping": undefined,
    "name": "pitotFailure",
    "permission": "rw",
    "type": 0,
    "value": 2929,
  },
  "transponderFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "transponder failure",
    "length": undefined,
    "mapping": undefined,
    "name": "transponderFailure",
    "permission": "rw",
    "type": 0,
    "value": 2927,
  },
  "turnCoordinatorFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "turn coordinator failure",
    "length": undefined,
    "mapping": undefined,
    "name": "turnCoordinatorFailure",
    "permission": "rw",
    "type": 0,
    "value": 2930,
  },
  "vacuumFailure": Offset {
    "category": "failure",
    "convert": "typeof {VAL} === \`boolean\` ? {VAL} ? 1 : 0 : !!{VAL}",
    "description": "vacuum failure",
    "length": undefined,
    "mapping": undefined,
    "name": "vacuumFailure",
    "permission": "rw",
    "type": 0,
    "value": 2931,
  },
}
`;
